aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwMultilabelRowCreator.scala

```diff
@@ -0,0 +1,361 @@
+package com.eharmony.aloha.dataset.vw.multilabel
+
+import com.eharmony.aloha.AlohaException
+import com.eharmony.aloha.dataset._
+import com.eharmony.aloha.dataset.density.Sparse
+import com.eharmony.aloha.dataset.vw.VwCovariateProducer
+import com.eharmony.aloha.dataset.vw.multilabel.json.VwMultilabeledJson
+import com.eharmony.aloha.dataset.vw.unlabeled.VwRowCreator
+import com.eharmony.aloha.reflect.RefInfo
+import com.eharmony.aloha.semantics.compiled.CompiledSemantics
+import com.eharmony.aloha.semantics.func.GenAggFunc
+import spray.json.JsValue
+
+import scala.collection.{breakOut, immutable => sci}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Created by ryan.deak on 9/13/17.
+  */
+final case class VwMultilabelRowCreator[-A, K](
+    allLabelsInTrainingSet: sci.IndexedSeq[K],
+    featuresFunction: FeatureExtractorFunction[A, Sparse],
+    defaultNamespace: List[Int],
+    namespaces: List[(String, List[Int])],
+    normalizer: Option[CharSequence => CharSequence],
+    positiveLabelsFunction: GenAggFunc[A, sci.IndexedSeq[K]],
+    classNs: Char,
+    dummyClassNs: Char,
+    includeZeroValues: Boolean = false
```

Comment:
I think we're missing something important here.  We have experimentally shown that when training on a very large multilabel problem the ability to randomly sample from the negative labels at train time both decreases train time and can increase model performance.  I think that a parameter to the row creator should be a function that allows me to sample a percentage of negative labels to add to each train line.  This needs to be random and the negative labels added needs to change with EVERY line.

Reply:
@jmorra I think the requirement that "*negative labels added need to change with EVERY line*" is important, but I don't think it fits into the current architecture.  It seems like you are (at least implicitly) advocating two things:

1. Intentionally making `apply` "*more*" impure, and definitely *not* idempotent or a function (in the math sense).
1. Maintaining state inside `VwMultilabelRowCreator` (by virtue of *(1)*)

What if we have a training set that is distributed across multiple nodes.  We apply the `VwMultilabelRowCreator` to the elements on each node and combine the results?  If we want to do what I think you are suggesting, I think we'll need to make some pretty big design changes.  I suggest the following:

1. We'll need a constructor `seed` parameter.  This could be a value, but if so, the calling code would be responsible for varying the seed for each node on which the `VwMultilabelRowCreator` is run.  It could also be an impure and side-effecting function is idempotent (in context), meaning it would produce the same result every time it is run on the same node.  This would allow the seed *generator* parameter to be set once and called on each node to produce the seed for that node. 
   ```scala
   seed: () => Long // Not necessarily a long but something to seed pseudo-randomness.
    ```
1. We'll need a new interface like `StatefulRowCreator`.  This interface will use something like the `State` monad in [cats](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/package.scala#L41-L42) or [scalaz](https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/package.scala#L147-L156).  I would recommend stack safety because its implementation stack-safe (remember the `tut` bug in 0.5.3?).
1. We'll have to use the `StatefulRowCreator` differently most likely, we could expose two methods, one that operates on single items and requires the caller the maintain the state and pass it in.  The other method could operate on multiple elements (collections, iterators) and hide the state information.

What do you think about this stuff?  I think if we try to jam this type of mutability into the current interface, it'll most likely produce disastrous results.

Reply:
@jmorra: I am proposing the following trait:

```scala
trait StatefulRowCreator[-A, +B, S] extends Serializable {
  // An initial state derived from the constructor
  def initialState: S

  // Caller maintains state.
  def apply(a: A, s: S): ((MissingAndErroneousFeatureInfo, Option[B]), S)

  // Iterator doesn't return state because it's non-strict.  Output Iterator length
  // is same as input iterator length.  May be infinite length.  This should be a constant
  // time operation.
  def apply(as: Iterator[A], s: S): Iterator[(MissingAndErroneousFeatureInfo, Option[B])]

  // Output Vector length is same as input length.  Returns the final state.
  // This is a linear time operation.
  def apply(as: Vector[A], s: S): (Vector[(MissingAndErroneousFeatureInfo, Option[B])], S)
}
```

I think it was a mistake in retrospect to not have `Option[B]` in the `RowCreator` trait.  This should be rectified here and we can address the `Option` in the `RowCreator` elsewhere.